src/Oscoin/Consensus/Nakamoto.hs

```diff
@@ -81,41 +100,71 @@ validateBlock prefix@(parent:_) blk
     ts = sinceEpoch . blockTimestamp . blockHeader
 
 validateBlock'
-    :: Serialise tx
-    => Block tx PoW              -- ^ Block to validate.
-    -> Either ValidationError () -- ^ Either a validation error, or success.
-validateBlock' Block{..}
-    | h <- blockDataHash blockHeader
-    , h /= hashTxs blockData =
+    :: ( Eq (Hash c)
+       , Serialise tx
+       , ByteArrayAccess (BlockHash c)
+       , Hashable c (BlockHeader c (Sealed c PoW))
+       , AuthTree.MerkleHash (Hash c)
+       )
+    => Block c tx (Sealed c PoW)     -- ^ Block to validate.
+    -> Either (ValidationError c) () -- ^ Either a validation error, or success.
+validateBlock' b
+    | h <- blockDataHash bHeader
+    , h /= hashTxs (blockData b) =
         Left $ InvalidDataHash h
-    | not (hasPoW blockHeader) =
-        Left $ InvalidBlockDifficulty (difficulty blockHeader)
-                                      (blockTargetDifficulty blockHeader)
+    | not (hasPoW bHeader) =
+        Left $ InvalidBlockDifficulty (difficulty bHeader)
+                                      (blockTargetDifficulty bHeader)
     | otherwise =
         Right ()
+  where
+      bHeader = blockHeader b
 
 mineNakamoto
-    :: forall m. Monad m
-    => (forall tx. [Block tx PoW] -> Difficulty)
-    -> Miner PoW m
-mineNakamoto difiFn getBlocks bh = do
+    :: forall c m.
+       ( Monad m
+       , ByteArrayAccess (BlockHash c)
+       , Hashable c (BlockHeader c (Sealed c PoW))
+       , Hashable c (BlockHeader c Unsealed)
+       )
+    => (forall tx. [Block c tx (Sealed c PoW)] -> Difficulty)
+    -> Miner c PoW m
+mineNakamoto difiFn getBlocks unsealedBlock = do
     blks <- getBlocks difficultyBlocks
-    go $ bh { blockTargetDifficulty = difiFn blks } $> PoW 0
+
+    let targetDifficulty = difiFn blks
+        bh = (blockHeader unsealedBlock) { blockTargetDifficulty = targetDifficulty }
+        candidateBlock = unsealedBlock { blockHeader = bh }
+
+    pure $ (`sealBlock` candidateBlock) <$> mine targetDifficulty (PoW 0)
+
   where
-    go :: BlockHeader PoW -> m (Maybe (BlockHeader PoW))
-    go hdr@BlockHeader { blockSeal = PoW nonce }
-        | hasPoW hdr            = pure $ Just $ hdr $> PoW nonce
-        | nonce < maxBound      = go hdr { blockSeal = PoW (nonce + 1) }
-        | otherwise             = pure Nothing
+    mine :: Difficulty -> PoW -> Maybe PoW
+    mine target (PoW nonce)
+        | blockDifficulty < target = Just (PoW nonce)
```

Comment:
I think it makes more sense to use `hasPoW` here like before, so we have only one place which verifies the seal.

Reply:
I think I couldn't use `hasPow` directly due to a type mismatch, but maybe now that the dust settled I can give this another spin.

Reply:
I think it's worth re-jiggling the implementation of `mineNakamoto` and/or `hasPoW` to make it work, if needed.

Reply:
cf21232